// @ohos/sdkmanager-common/build/src/core/task/abstract-component-task.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractComponentTask = void 0;
const path_1 = __importDefault(require("path"));
const tempDir = '.temp';
class AbstractComponentTask {
    constructor(component) {
        this.component = component;
    }
    execute(progress) {
        return __awaiter(this, void 0, void 0, function* () {
            progress.info(`Install task started: ${this.getComponent().getDisplayName()} ${this.getComponent().getVersion()}`);
            yield this.prepare(progress);
            yield this.complete(progress);
            progress.info(`Install task finished: ${this.getComponent().getDisplayName()} ${this.getComponent().getVersion()}`);
            return Promise.resolve();
        });
    }
    getRootWorkDir() {
        return path_1.default.resolve(this.getSdkRoot(), tempDir);
    }
    getWorkDir() {
        return path_1.default.resolve(this.getRootWorkDir(), this.component.getPath(), this.component.getVersion());
    }
    prepare(progress) {
        progress.debug('prepare');
        return Promise.resolve(undefined);
    }
    complete(progress) {
        progress.debug('complete');
        return Promise.resolve(undefined);
    }
}
exports.AbstractComponentTask = AbstractComponentTask;
