// @ohos/sdkmanager-common/build/src/core/util/os-type-util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEnvPathSeparator = exports.getEnvPathVariable = exports.getOsType = exports.isMac = exports.isLinux = exports.isWindows = void 0;
const node_os_1 = __importDefault(require("node:os"));
const path_1 = __importDefault(require("path"));
function isWindows() {
    return node_os_1.default.platform() === 'win32';
}
exports.isWindows = isWindows;
function isLinux() {
    return node_os_1.default.platform() === 'linux';
}
exports.isLinux = isLinux;
function isMac() {
    return node_os_1.default.platform() === 'darwin';
}
exports.isMac = isMac;
const getOsType = () => {
    if (isLinux()) {
        return 'linux';
    }
    else if (isWindows()) {
        return 'windows';
    }
    else if (isMac()) {
        return 'darwin';
    }
    else {
        return 'other';
    }
};
exports.getOsType = getOsType;
function getEnvPathVariable() {
    return isWindows() ? 'Path' : 'PATH';
}
exports.getEnvPathVariable = getEnvPathVariable;
function getEnvPathSeparator() {
    return path_1.default.delimiter;
}
exports.getEnvPathSeparator = getEnvPathSeparator;
